core/src/bytecode/ast/compat.rs

```diff
@@ -791,13 +831,111 @@ impl<'ast> FromAst<Annotation<'ast>> for term::TypeAnnotation {
     }
 }
 
-impl<'ast> FromAst<record::Field<'ast>> for term::record::Field {
-    fn from_ast(field: &record::Field<'ast>) -> Self {
-        term::record::Field {
-            value: field.value.as_ref().map(|term| term.to_mainline()),
+impl<'ast> FromAst<StrChunk<Ast<'ast>>> for term::StrChunk<term::RichTerm> {
+    fn from_ast(chunk: &StrChunk<Ast<'ast>>) -> Self {
+        match chunk {
+            StrChunk::Literal(s) => term::StrChunk::Literal(s.clone()),
+            StrChunk::Expr(expr, indent) => term::StrChunk::Expr(expr.to_mainline(), *indent),
+        }
+    }
+}
+
+/// Similar to [record::FieldPathElem], but for the mainline representation: either an identifier
+/// or a quoted identifier.
+pub enum FieldName {
+    Ident(LocIdent),
+    Expr(Vec<StrChunk<term::RichTerm>>, TermPos),
+}
+
+impl FromAst<record::FieldPathElem<'_>> for FieldName {
+    fn from_ast(elem: &record::FieldPathElem<'_>) -> Self {
+        match elem {
+            record::FieldPathElem::Ident(id) => FieldName::Ident(*id),
+            record::FieldPathElem::Expr(chunks, pos) => {
+                let chunks = chunks.iter().map(ToMainline::to_mainline).collect();
+                FieldName::Expr(chunks, *pos)
+            }
+        }
+    }
+}
+
+impl<'ast> FromAst<record::FieldDef<'ast>> for (FieldName, term::record::Field) {
+    fn from_ast(field: &record::FieldDef<'ast>) -> Self {
```

Comment:
This is mostly `parser::utils::build_record` adapted to take a new AST as an input.

Reply:
Sorry, this is rather `elaborate_field_def`.